### ISSUE TITLE
Feature: Reproducible per-workspace/per-pipeline Python environments using uv

### ISSUE THREAD

**Issue by LouisDeconinck** [slothflowlabs/duckle#246]
## The problem

A first-class Python stage is only useful in production if its dependencies are reproducible, see #245 

Data pipelines commonly require packages such as:

- pyarrow
- polars
- lxml
- beautifulsoup4
- pymupdf
- docling
- splink
- scikit-learn
- lightgbm
- xgboost

Relying on whatever happens to be installed in the system Python environment makes a pipeline difficult to reproduce between:

- developer laptops
- CI
- headless Duckle runners
- production servers
- AI-agent execution environments

## Proposed solution

Add optional Python environment management based on `uv`.

Duckle should support a workspace or pipeline containing:

```text
pyproject.toml
uv.lock
````

A `code.python` stage would execute in that locked environment.

Suggested behavior:

1. Detect `pyproject.toml` / `uv.lock`.
2. Create/cache the environment automatically.
3. Reuse the environment until the lockfile changes.
4. Include the dependency-lock hash in run metadata.
5. Fail clearly if dependency resolution fails.
6. Support offline execution when packages are already cached.
7. Make the same environment available to desktop, CLI, CI and MCP runs.

Optional per-node dependencies could also be supported:

```yaml
python:
  dependencies:
    - splink==...
    - pyarrow==...
```

but a workspace-level `pyproject.toml` + `uv.lock` would probably be simpler and more Git-friendly.

## Reproducibility

The run manifest should ideally record:

* Python version
* `uv.lock` hash
* platform
* dependency environment hash

This would make Python-based Duckle runs reproducible in the same spirit as Duckle's existing signed run manifests.

## Build/export behavior

When Duckle builds a deployable pipeline artifact, it should either:

* package the environment,
* package/install the lockfile,
* or fail clearly if the target runtime does not contain the required environment.

## Alternatives / workarounds

Users can manually maintain virtualenvs today and call Python from shell stages, but this makes deployment and agent-driven execution substantially less deterministic.

**Comment by SouravRoy-ETL** [comment]
Landed on `main` in 18ed0f7.

A virtual environment at `.venv` in the workspace is now used by `code.python` on every machine - laptop, CI and the headless runner alike. `uv venv` and the stdlib `venv` module both produce that layout, so this works with uv without depending on it. `DUCKLE_PYTHON_BIN` still overrides it.

Deliberately not included: Duckle does not create or sync the environment for you. That would mean a network install at pipeline run time, which contradicts running on your own infrastructure and turns a missing package into a mid-run download. `uv venv && uv pip install ...` stays your step; Duckle just finds what it produced.

This also matters for #245: `transform(table)` needs pyarrow, and now it needs it in the environment that travels with the pipeline rather than whatever the machine happens to have.

**Comment by LouisDeconinck** [comment]
Using the workspace .venv is a useful first step, but I don't think this fully provides the reproducibility #246 was aiming for yet.

At the moment Duckle selects the interpreter from .venv, but it does not verify how that environment was created or whether it corresponds to the workspace's pyproject.toml / uv.lock.

That means:

machine A:
    .venv with splink==X

machine B:
    .venv with splink==Y

can run the same Duckle pipeline with different results while both appear valid.

Could Duckle add a lightweight uv lock integration without compromising air-gapped runtime behavior?

For example:

1. Detect pyproject.toml + uv.lock.
2. Provide an explicit prepare/sync command, rather than installing packages during a normal pipeline run:
       duckle python prepare
   or equivalent.
3. Cache/reuse .venv until uv.lock changes.
4. Record at least:
       Python version
       uv.lock SHA-256
       platform
       environment/lock hash
   in the signed run manifest.
5. On CI/server execution, optionally fail clearly when .venv does not correspond to the current lock.
6. When building/deploying a pipeline containing code.python, either package/prepare the environment or clearly report that it must be prepared on the target.

This preserves the nice "nothing is fetched at runtime" property while giving us actual reproducibility between developer, CI, MCP agent and production runner.

**Comment by SouravRoy-ETL** [comment]
Agreed that finding the interpreter is not the same as pinning the environment, and your machine A / machine B example is the right way to put it - both look valid and produce different answers.

Your proposed split is also the part I would keep: an explicit prepare/sync step, never an install during a normal pipeline run. Installing at run time is what would break the air-gapped and scheduled cases, so a separate command that can be run once, in CI or during provisioning, is the only shape that fits.

Detecting `pyproject.toml` + `uv.lock` and verifying the `.venv` matches them is the useful half - reporting a drifted environment at plan time would catch exactly your scenario before it produces different results rather than after.

**Comment by SouravRoy-ETL** [comment]
On `main` in `62a0bed`. Your six points, in order.

## 1. Detect `pyproject.toml` + `uv.lock`

The lock is the opt-in. With none committed, nothing changes and `code.python` resolves the interpreter exactly as it did.

## 2. An explicit prepare/sync, never an install during a run

```bash
duckle-runner python prepare    # uv sync --frozen in the workspace
duckle-runner python check      # exit 1 when the environment contradicts the lock
```

`prepare` is a provisioning step. Nothing is fetched during a pipeline run, which is the property you asked to keep.

## 3. Reuse `.venv` until the lock changes

Better than a stamp, and this is the part I want your eyes on. Duckle does not record what `prepare` intended - it reads **what is actually installed**, from the `*.dist-info` directory names in `site-packages`. No file is opened and no interpreter is started, so it is cheap enough to run before every pipeline that has a Python stage.

That means a `.venv` built by `uv sync`, by `pip install`, or by hand is judged the same way: by its contents. A stamp file records an intention; `dist-info` records the fact. It also means your machine A / machine B case is caught even when both machines ran `prepare` correctly at different times.

## 4. Recorded in the signed manifest

```json
"pythonEnvironment": {
  "pythonVersion": "3.12.4",
  "platform": "windows-x86_64",
  "lockSha256": "93171dbc9ecb...",
  "environmentHash": "6b4f40d0e5a2...",
  "packageCount": 41
}
```

Only for a pipeline that actually has a `code.python` node - an unused interpreter in every manifest would be noise. `environmentHash` is SHA-256 over the sorted normalised `name==version` set plus the Python version, so two machines that installed the same things in a different order agree and one different patch version does not.

## 5. Fail clearly when `.venv` does not correspond to the lock

Live, refusing before anything executes:

```
error : config: python: the workspace .venv is not the environment uv.lock
        describes, so this run would not be the run the lock says it is:
          splink: installed 3.9.1, locked 4.0.0

        Run `duckle-runner python prepare` to rebuild it from the lock, or set
        DUCKLE_PYTHON_ALLOW_DRIFT=1 to run anyway.
```

Two shapes fail: **a version that contradicts the lock**, and **a package the lock never mentions** (something installed on top). One shape reports and does not fail: **a locked package that is not installed.**

That last one is a deliberate call and the place I could be wrong. `uv.lock` carries the whole resolution across platforms and extras, and honouring its markers properly means implementing PEP 508 marker evaluation. Treating every absent package as an error would fail correct runs on any platform-specific dependency - `pywin32` on Linux, and so on. A package that really is needed and really is missing raises `ImportError` on the first row, which is already unambiguous. If you would rather have strict marker evaluation, that is a real piece of work and worth its own issue.

`DUCKLE_PYTHON_ALLOW_DRIFT=1` downgrades the refusal to a warning. There is always a machine where the rule is wrong, and a check with no way past it gets deleted rather than fixed.

## 6. Build/deploy

Not done. `python check` gives a deploy pipeline the gate today - it exits 1 - but the bundle exporter does not yet package or verify the environment on the target. That is the honest state of it.

## Also

This closed the last item on #252: a cached `code.python` output is keyed on the environment hash, and a stage with no workspace environment of its own is **not cached at all** - it runs against whatever interpreter the machine has, which can change without anything noticing, so there is no identity to key on.

## Tests

Eight in `crates/duckdb-engine/src/pyenv.rs`, covering PEP 503 name normalisation, the `name-version` split for a hyphenated package name, all three drift shapes, no-lock-means-no-check, and that the environment hash is contents rather than order and includes the Python version.

**Comment by LouisDeconinck** [comment]
`62a0bed` solves the runtime reproducibility problem I was after, and the `2f80200` follow-up also removes an important false-positive for normal stdlib-created virtual environments.

From my original six points, I think only the **build/deploy handoff** remains.

I don't think Duckle needs to package an entire Python environment. A simpler contract would be enough:

* a built/deployed pipeline containing `code.python` carries the lock contract;
* the target environment is verified before execution; and
* if it is missing or does not match, Duckle clearly requires `duckle-runner python prepare` (or an equivalent deploy-time preparation step).

The important property is that a deployable Python pipeline cannot silently run against an arbitrary or unprepared target interpreter.

Once that target/deployment preflight exists, I would consider #246 complete.

**Comment by SouravRoy-ETL** [comment]
The deployment preflight is on `main` in `a9b55e3`. Your contract, in both halves.

## The bundle carries the lock

A build whose pipeline has a `code.python` step now ships `uv.lock` and `pyproject.toml` in the artifact:

```
entries: 6
  manifest.json                          654
  pipeline/csv_split_child.pipeline.json 1707
  pyproject.toml                          45
  uv.lock                                 66
```

The **lock**, not the environment - exactly as you framed it. Duckle does not package wheels, because resolving dependencies at run time is what an air-gapped or scheduled box cannot have. A build with a Python step and no lock says so on stderr rather than shipping something whose Python behaviour the target decides.

## The target is verified before execution

And here I owe you a correction, because what I shipped in `62a0bed` did **not** have this property. My own rule that a locked-but-not-installed package only reports and does not fail - right on its own, since a lock resolves across platforms - left the hole exactly where you were pointing: a target that was never prepared has an absent or empty `.venv`, produces **nothing but** Missing entries, and so ran happily against whatever interpreter the machine had.

Refused now, before the run starts:

```
error : config: python: /srv/app declares a locked environment (uv.lock) but
        nothing is installed in .venv, so this target is not prepared. Run
        `duckle-runner python prepare` on this machine before the pipeline runs,
        or point DUCKLE_PYTHON_BIN at the interpreter you mean. Without one of
        those the run would use whatever Python the machine happens to have,
        which is what committing a lock file exists to rule out.
```

The trigger is **nothing installed**, not **no interpreter**: it is true whether `.venv` is absent or merely never synced, and it does not depend on where a platform puts the binary.

`DUCKLE_PYTHON_BIN` is exempt. Naming an interpreter explicitly is a decision, not an unprepared machine, and refusing there would have broken a legitimate setup.

Verified by running it, not by reading it: an unprepared workspace is refused, the same workspace with `DUCKLE_PYTHON_BIN` set runs, and the built bundle's ZIP entries are listed above.

That is the property you asked for - a deployable Python pipeline cannot silently run against an arbitrary or unprepared target interpreter - so from my side #246 is complete.
